### DISCUSSION TITLE
CBOR: $type is missing?

### DISCUSSION THREAD

**Discussion by dongle-the-gadget** [bluesky-social/atproto#1290] [Dev Debugging] [answered]
I'm using PeterO/CBOR to parse the CBOR for `com.atproto.sync.subscribeRepos` in C# (the CBOR implementation isn't DAG-CBOR, however if I understood correctly there should be interoperability between both). However, the CBOR payloads the Bluesky API returned didn't have `$type`, but rather a `t` property with the type info I'm looking for (see picture below).

<img width="444" alt="image" src="https://github.com/bluesky-social/atproto/assets/29563098/e693cee2-aca3-4aee-86d5-93cfea9fd958">

I'm not sure if I'm missing anything, or is this expected?

**Comment by myConsciousness** [comment]
Hi @dongle-the-gadget ,

The `t` stands for the `$type` of this response. So it means `#commit` structure below.

https://github.com/bluesky-social/atproto/blob/7e0201d4d6d1a69fe255b6643f3d0b745d2ee9e9/lexicons/com/atproto/sync/subscribeRepos.json#L34-L61

**Reply by dongle-the-gadget** [reply]
Does that mean `$type` will be represented as `t` in CBOR objects?

**Comment by bnewbold** [comment, marked as answer]
Check out the Event Stream specs, they talks about the use of `t` in DAG-CBOR objects returned in the frame header DAG-CBOR object: https://atproto.com/specs/event-stream#framing

In short, the `t` field indicates the `union` type variant of the payload, as a short ref (like `#commit`), not a fully-qualified type path like a `$type` field would have.

**Reply by dongle-the-gadget** [reply]
Right. However, if there were a `union` type inside the content, would it use `t` or `$type`?

**Reply by bnewbold** [reply]
If there was a nested `union` inside the body payload, it would have `$type`, in both JSON and DAG-CBOR encodings. As another example, top-level records always have a `$type`, and field name is `$type` in both JSON and DAG-CBOR.

The Event Stream header use of `t` in DAG-CBOR, and having the truncated union variant name, is an exception, to make the wire protocol more efficient. The header is not a regular Lexicon object, which is why this is acceptable. The payload/body of the event frame *is* a regular Lexicon object.

**Reply by dongle-the-gadget** [reply]
Thanks for the explanation.

**Comment by sugyan** [comment]
related discussion: https://github.com/bluesky-social/atproto/discussions/1137
